core/solver/idr.cpp

```diff
@@ -0,0 +1,261 @@
+
+namespace gko {
+namespace solver {
+namespace idr {
+
+
+GKO_REGISTER_OPERATION(initialize, idr::initialize);
+GKO_REGISTER_OPERATION(step_1, idr::step_1);
+GKO_REGISTER_OPERATION(step_2, idr::step_2);
+GKO_REGISTER_OPERATION(step_3, idr::step_3);
+GKO_REGISTER_OPERATION(compute_omega, idr::compute_omega);
+GKO_REGISTER_OPERATION(fill_array, components::fill_array);
+
+
+}  // namespace idr
+
+
+template <typename ValueType>
+std::unique_ptr<LinOp> Idr<ValueType>::transpose() const
+{
+    return build()
+        .with_generated_preconditioner(
+            share(as<Transposable>(this->get_preconditioner())->transpose()))
+        .with_criteria(this->stop_criterion_factory_)
+        .on(this->get_executor())
+        ->generate(
+            share(as<Transposable>(this->get_system_matrix())->transpose()));
+}
+
+
+template <typename ValueType>
+std::unique_ptr<LinOp> Idr<ValueType>::conj_transpose() const
+{
+    return build()
+        .with_generated_preconditioner(share(
+            as<Transposable>(this->get_preconditioner())->conj_transpose()))
+        .with_criteria(this->stop_criterion_factory_)
+        .on(this->get_executor())
+        ->generate(share(
+            as<Transposable>(this->get_system_matrix())->conj_transpose()));
+}
+
+
+template <typename ValueType>
+void Idr<ValueType>::apply_impl(const LinOp *b, LinOp *x) const
+{
+    using std::swap;
+    using Vector = matrix::Dense<ValueType>;
+    using NormVector = matrix::Dense<remove_complex<ValueType>>;
+
+    auto exec = this->get_executor();
+
+    auto one_op = initialize<Vector>({one<ValueType>()}, exec);
+    auto neg_one_op = initialize<Vector>({-one<ValueType>()}, exec);
+
+    auto dense_b = as<Vector>(b);
+    auto dense_x = as<Vector>(x);
+
+    constexpr uint8 RelativeStoppingId{1};
+
+    const auto problem_size = system_matrix_->get_size()[0];
+    const auto nrhs = dense_b->get_size()[1];
+
+    auto residual = Vector::create_with_config_of(dense_b);
+    auto v = Vector::create_with_config_of(dense_b);
+    auto t = Vector::create_with_config_of(dense_b);
+    auto helper = Vector::create_with_config_of(dense_b);
+
+    auto m =
+        Vector::create(exec, gko::dim<2>{subspace_dim_, subspace_dim_ * nrhs});
+
+    auto g =
+        Vector::create(exec, gko::dim<2>{problem_size, subspace_dim_ * nrhs});
+    auto u =
+        Vector::create(exec, gko::dim<2>{problem_size, subspace_dim_ * nrhs});
+
+    auto f = Vector::create(exec, gko::dim<2>{subspace_dim_, nrhs});
+    auto c = Vector::create(exec, gko::dim<2>{subspace_dim_, nrhs});
+
+    auto omega = Vector::create(exec, gko::dim<2>{1, nrhs});
+    auto residual_norm = NormVector::create(exec, dim<2>{1, nrhs});
+    auto tht = Vector::create(exec, dim<2>{1, nrhs});
+    auto t_norm = NormVector::create(exec, dim<2>{1, nrhs});
+    auto alpha = Vector::create(exec, gko::dim<2>{1, nrhs});
+
+    bool one_changed{};
+    Array<stopping_status> stop_status(exec, nrhs);
+
+    auto subspace_vectors_ =
```

Comment:
subspace_vectors is col-major, right? Could you add some description about it because we usually use row-major?